lib/rubocop/cop/style/redundant_begin.rb

```diff
@@ -28,14 +28,24 @@ class RedundantBegin < Cop
         MSG = 'Redundant `begin` block detected.'.freeze
 
         def on_def(node)
-          return unless node.body && node.body.kwbegin_type?
-
-          add_offense(node.body, location: :begin)
+          check(node)
         end
         alias on_defs on_def
 
+        def on_block(node)
+          return if target_ruby_version < 2.5
+          return if node.braces?
+          check(node)
+        end
+
         private
 
+        def check(node)
+          return unless node.body && node.body.kwbegin_type?
```

Comment:
I think the cop is not accounting for the case where there are additional expressions before the `begin`. In those cases the `body` is wrapped in a `begin` node. Consider:

```ruby
foo do
  bar

  begin
    baz
  rescue
    qux
  end
end
```

Which will result in the following AST:

```
(block
  (send nil :foo)
  (args)
  (begin
    (send nil :bar)
    (kwbegin
      (rescue
        (send nil :baz)
        (resbody nil nil
          (send nil :qux)) nil))))
```

Reply:
There's a difference between `begin` and `kwbegin`. The code seems correct to me, although adding an extra test case probably won't hurt.

Reply:
The problem is in how parser handles most bodies. If the body a single expression, it omits wrapping it in a `begin`. So if we want to check for offenses in both single- and multiple expression blocks, we need to do something like:

```ruby
if body.offending_type? || body.begin_type? && body.children.any?(&:offending_type?)
```

The code above won't register an offense currently, even though it can be changed to:

```ruby
foo do
  bar
  baz
rescue
  qux
end
```

in Ruby 2.5.

Reply:
> The code above won't register an offense currently, even though it can be changed to:

I think the changing is not safe. The code has different behaviour when `bar` raises an error.
In some cases the changing is safe, but in other cases the changing is not safe.
For example:

```ruby
# In this case, omitting `begin` and `end` is safe.
# Because `Repository.find` (`ActiveRecord method`) never raise `Octokit` error.
foo do
   repository = Repository.find(params[:id])
  begin
    Octokit::Client.new.repository(repository.full_name)
  rescue Octokit::NotFound
    handle_error
  end
end

# In this case, omitting `begin` and `end` is not safe.
# Because `Repository.find` may raise `ActiveRecord::RecordNotFound` error.
foo do
   repository = Repository.find(params[:id])
  begin
    user = User.find(params[:user_id])
  rescue ActiveRecord::RecordNotFound
    handle_error
  end
end
```

It depends on an exception and methods, so I think this cop should not add any offense for the code.